src/west/manifest.py

```diff
@@ -537,14 +573,72 @@ def _load_project(self, pdata):
         except ValueError as ve:
             self._malformed(ve.args[0])
 
+        import_files = self._imported_files(project, import_key)
+
         # The name "manifest" cannot be used as a project name; it
         # is reserved to refer to the manifest repository itself
         # (e.g. from "west list"). Note that this has not always
         # been enforced, but it is part of the documentation.
         if project.name == 'manifest':
             self._malformed('no project can be named "manifest"')
 
-        return project
+        return project, import_files
+
+    def _imported_files(self, project, import_key):
+        ktype = type(import_key)
+
+        if not import_key or (isinstance(project, ManifestProject) and
+                              ktype == bool):
+            return []
+
+        if ktype == bool:
+            ret = os.path.join(project.abspath, _WEST_YML)
+            if not os.path.isfile(ret):
+                _file_not_found(ret)
+            return [ret]
+        elif ktype == str:
+            path = os.path.join(project.abspath, import_key)
+            if os.path.isfile(path):
+                # Take the user's word for it, even if that doesn't end in
+                # something from _YML_EXTS.
+                return [path]
+            elif os.path.isdir(path):
+                # Find the YAML files in the directory. It's OK if there
+                # aren't any.
+                ret = []
+                for f in os.listdir(path):
+                    if os.path.splitext(f) in _YML_EXTS:
+                        ret.append(f)
+                return ret
+            else:
+                _file_not_found(f)
+        elif ktype == list:
+            return list(itertools.chain(self._imported_files(project, k)
+                                        for k in import_key))
+        elif ktype == dict:     # TODO
+            raise NotImplementedError('import: <map> is unimplemented')
+        else:
+            self._malformed('{}: invalid import key {} (type {})'.
+                            format(project.name, import_key, ktype))
+
+    def _import(self, project, file):
+        try:
+            data = project.file_contents_at(file)
+        except subprocess.CalledProcessError:
+            # Fall back on the importer.
+            data = self._importer(project, file, project.revision)
```

Comment:
> Do we assume that _importer() will actually update the current working tree or just fetch into the local refspace and then git show? 

At this level we make no assumptions about how the data are retrieved. We might not involve the filesystem at all. Try this:

```
$ wget -q -O- https://raw.github.com/zephyrproject-rtos/zephyr/v1.14.1/west.yml
```

This enables doing imports with `Manifest.from_data()` and not just `Manifest.from_file()`. We don't need to assume an installation exists anywhere on disk to parse the manifest data.

Reply:
> Also, should this really be called `importer`? This is not really importing anything but actually "fetching" a particular file at a particular revision. Should it be called `fetcher`?

It's the object that is used as a fallback from the `_import()` method when the default `git show` fails. It isn't necessarily fetching anything over the network -- the data could be coming from somewhere else on the file system, if the user of this API has some special knowledge of where to fetch the data from. (Remember that this API is callable from extensions too.)

The in-tree user of this API is `west update`. That code might be running before the repository is even cloned. If that's the case, I think the right thing to do is to clone the new repository into a temporary subdirectory of .west, because we don't know where this project should go yet -- the "parent" manifest might override its `path:`. Only when all the data are available and the manifest is completely resolved can `west update` move newly-cloned repositories into their final locations.